### PR TITLE
Added handler for Razor Endpoint

### PR DESCRIPTION
**Pull request by marcarro**
# Summary of changes
Created handler corresponding to a new LSP request endpoint in razor, to be used in Extract To Component project.

### CODE REVIEW

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public required TextDocumentIdentifier Project { get; init; }
+
+    [JsonPropertyName("hostDocumentVersion")]
+    public required int HostDocumentVersion { get; init; }
+
+    [JsonPropertyName("newContents")]
+    public required string NewContents { get; init; }
+}
+
+// Not sure where to put these two records
```
Comment (marcarro) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:30]: three *

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    public async Task<RazorComponentInfo?> HandleRequestAsync(RazorComponentInfoParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        var project = context.Solution?.GetProject(request.Project);
+
+        if (project is null)
+        {
+            return null;
+        }
+
+        var document = context.Solution.GetDocument(request.Document);
```
Comment (ryzngard) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:55]: If you're getting a document you can get the project it is in directly

```suggestion
        var document = context.Solution.GetDocument(request.Document);
        var project = document.Project;
```

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = semanticModel.SyntaxTree;
+        var root = (CompilationUnitSyntax)syntaxTree.GetRoot(cancellationToken);
```
Comment (ryzngard) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:58]: I don't think the cast is needed here

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            Fields = fields
+        };
+
+        return result;
+    }
+
+    private static (List<MethodInsideRazorElementInfo>, List<SymbolInsideRazorElementInfo>) GetInfoInsideRazorDocument(CompilationUnitSyntax root, SemanticModel semanticModel, CancellationToken cancellationToken)
+    {
+        var invocationExpressions = root.DescendantNodes().OfType<InvocationExpressionSyntax>().ToList();
+        var identifierNames = root.DescendantNodes().OfType<IdentifierNameSyntax>().ToList();
```
Comment (ryzngard) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:74]: This looks like you're getting all the invocations and identifiers in the whole file, not just ones used by razor.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        
+        List<MethodInsideRazorElementInfo> methods = [];
```
Comment (ryzngard) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:76]: Likely these should be pooled types

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        List<SymbolInsideRazorElementInfo> fields = [];
+        
+        foreach (var invocation in invocationExpressions)
+        {
+            var invocationOperation = semanticModel.GetOperation(invocation, cancellationToken) as IInvocationOperation;
+            var invocationDataFlow = semanticModel.AnalyzeDataFlow(invocation);
```
Comment (ryzngard) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:82]: This appears to be unused

**Review by ryzngard — COMMENTED** [review]
Some initial feedback. It's hard to follow how this information will be used. Do you have the razor side ready? It would be helpful to see that side to better comment on this code.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        _globalOptions = globalOptions;
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<RazorComponentInfo?> HandleRequestAsync(RazorComponentInfoParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        var project = context.Solution?.GetProject(request.Project);
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:48]: I don't know what is in `request.Project`, but generally speaking the way things work in LSP would be to send the Uri of the document in the request, and then access everything via the `context` parameter. So `var document = context.GetRequiredDocument();`, and from there if you need the project, it would just be `document.Project`.

Seems like this project isn't needed though.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs (hunk 4 above)
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:74]: Rather that descending through the whole tree twice, is it better to do it once and look for both of the node types you care about?

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs (hunk 6 above)
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:82]: Can you include a comment on what this is analyzing? Is it the parameters to the method, or the method itself?

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        foreach (var identifier in identifierNames)
+        {
+            var symbolInfo = semanticModel.GetSymbolInfo(identifier, cancellationToken);
+            if (symbolInfo.Symbol is IFieldSymbol or IPropertySymbol)
+            {
+                var field = new SymbolInsideRazorElementInfo
+                {
+                    Name = symbolInfo.Symbol.Name,
+                    Type = symbolInfo.Symbol.GetType().ToString()
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:124]: Pretty sure this is going to end up being `Microsoft.CodeAnalysis.SourceFieldSymbol` or something. Unlikely to be what you intended.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                };
+
+                fields.Add(field);
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:127]: Seems like there should be some filtering of duplicates here, if a field is referred to more than once

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            return null;
+        }
+
+        var document = context.Solution.GetDocument(request.Document);
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = semanticModel.SyntaxTree;
+        var root = (CompilationUnitSyntax)syntaxTree.GetRoot(cancellationToken);
+
+        var (methods, fields) = GetInfoInsideRazorDocument(root, semanticModel, cancellationToken);
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:60]: It would be good to have a comment somewhere about what this endpoint actually does and is used for? I don't know why Razor wants to know about all of the method calls and fields used in a C# file.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            var parameterTypes = targetMethod.GetParameters().Select(parameter => parameter.Type.ToDisplayString()).ToList();
+            if (parameterTypes is null)
+            {
+                continue;
+            }
+
+            var methodInfo = new MethodInsideRazorElementInfo
+            {
+                Name = targetMethod.Name,
+                ReturnType = operationReturnType.ToNameDisplayString(),
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:109]: What happens with this string? I don't think `ToNameDisplayString()`, or `ToDisplayString()` used above, guarantee that the result will be valid C# code.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(RazorComponentInfoHandler)), Shared]
+[Method(RazorComponentInfoName)]
+internal sealed class RazorComponentInfoHandler : ILspServiceRequestHandler<RazorComponentInfoParams, RazorComponentInfo?>
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:30]: Given that this handler seems to deal with a specific document, it should probably be a `ILspServiceDocumentRequestHandler`, to get the value out of the `context` parameter I mentioned below.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    public required TextDocumentIdentifier NewDocument { get; init; }
+
+    [JsonPropertyName("project")]
+    public required TextDocumentIdentifier Project { get; init; }
+
+    [JsonPropertyName("hostDocumentVersion")]
+    public required int HostDocumentVersion { get; init; }
+
+    [JsonPropertyName("newContents")]
+    public required string NewContents { get; init; }
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:27]: What goes in here?

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+using Roslyn.LanguageServer.Protocol;
+
+internal sealed record RazorComponentInfoParams
+
+{
+    [JsonPropertyName("document")]
+    public required TextDocumentIdentifier Document { get; init; }
+
+    [JsonPropertyName("newDocument")]
+    public required TextDocumentIdentifier NewDocument { get; init; }
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:18]: What goes in here?

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+
+internal sealed record MethodInsideRazorElementInfo
+{
+    public required string Name { get; set; }
+
+    public required string ReturnType { get; set; }
+
+    public required List<string> ParameterTypes { get; set; }
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:44]: Since these types are all for serialization, the `List` here, and the two in `RazorComponentInfo` should probably just be arrays.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+    [JsonPropertyName("hostDocumentVersion")]
+    public required int HostDocumentVersion { get; init; }
+
+    [JsonPropertyName("newContents")]
+    public required string NewContents { get; init; }
+}
+
+// Not sure where to put these two records
+internal sealed record RazorComponentInfo
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:31]: This doesn't seem to have anything to do with Razor components. Doesn't this contain C# info, not Razor info?

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs (hunk 1 above)
Comment (davidwengier) [reply src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoParams.cs:30]: I think here is fine

**Review by davidwengier — COMMENTED** [review]
I'm not sure what the aim of this endpoint is, so some of my comments might be irrelevant. Feel free to ignore :)

File: src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs (hunk 2 above)
Comment (marcarro) [reply src/LanguageServer/Protocol/ExternalAccess/Razor/RazorComponentInfoHandler.cs:55]: Added some more information in the initial post!

**Review by davidwengier — COMMENTED** [review]
I think there are two areas to focus on here: Firstly, generally speaking this is pretty clean code, and I can see what it's doing. What is not clear is _why_ it is doing it, and that is where code comments would be helpful.

Secondly, C# is very complicated, and it's usually a worrying sign to take shortcuts and make assumptions. eg, assuming that the first token of something will be important is unlikely to hold in all cases. Assuming that a matching on strings, names, `ToDisplayString()` etc. will work is similarly unlikely to hold in all cases. Roslyn is very powerful, and when things get reduced to comparing strings, it usually means that power is being thrown away.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+        var isWrittenTo = writtenInsideBlock.Any(symbol => symbol == declarationInfo.Name);
+
+        // Handle special case: attribute is string type or value type.
+        // Attributes of these types are not added to the 'WrittenInside' property of a data flow analysis when written to or mutated.
+
+        // Erring on the side of caution, assume they are written to if they are involved in some type of expression.
+
+        // The 'isWrittenTo' property is not critical to functionality in current usage; it's only used in ExtractToComponent
+        // to determine if a code attribute that has been promoted to a parameter in a component should include a comment warning.
+        if (typeInfo.Type.ToDisplayString() == "string" || typeInfo.Type.IsValueType)
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:196]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhGAzAATZkDKALlAJYB2A5oQN6FlcXm0MsAKDDgAMZBAEoybMgF9C8ogVIUsZAMLTO3FRhRkAsgKkcC3cxRwA6Pk2biyAXjIAiAJIByALZkAhpZEXAG5tCzJmABsIYF8IkBBhGzo7MgAvJ1dPH1iAZwg/AODQi1sWMgBPDMYYAHdqZMF3bzJGCBoAQhcJELNuRVkgA==

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+        return type.ToString();
+    }
+
+    private static string FormatType(ITypeSymbol typeSymbol)
+    {
+        // Check if the symbol is a named type symbol (e.g., List<T>)
+        if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+        {
+            // Get the base name of the type, e.g., "List"
+            var typeName = namedTypeSymbol.Name;
```
Comment (marcarro) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:290]: The case with integer bothers me a bit: "int" will become "Int32" (etc.) after accessing the type symbol name. This is technically correct and causes no errors, but I wonder if there's a way to just keep the type name as the user defined it.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+        var generatedSpans = request.GeneratedDocumentRanges.Select(r => ProtocolConversions.RangeToTextSpan(r, sourceText));
+
+        // First, get the class declaration for the component (implements Microsoft.AspNetCore.Components.ComponentBase). There might be a better way to get the type.
+        var componentBaseSymbol = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Components.ComponentBase");
+        if (componentBaseSymbol is null)
+        {
+            return null;
+        }
+
+        var classDeclarationNode = root.DescendantNodes().OfType<ClassDeclarationSyntax>().FirstOrDefault(classSymbol => InheritsFromComponentBase(componentBaseSymbol, classSymbol, semanticModel));
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:83]: Nit: The parameter is not a `classSymbol` it is a `classNode`. "Symbol" has a specific meaning in Roslyn, in that it comes from the semantic model not the syntax tree.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        var fieldsInClass = fieldsBuilder.ToImmutableAndFree();
+        var propertiesInClass = propertiesBuilder.ToImmutableAndFree();
+
+        var identifiersInRange = identifiersInClass.Where(identifier => generatedSpans.Any(span => span.Contains(identifier.Span)))
+                    .Select(identifier => new IdentifierAndSymbol
+                    {
+                        Identifier = identifier,
+                        Symbol = semanticModel.GetSymbolInfo(identifier).Symbol
+                    })
+                    .Where(x => x.Symbol is not null);
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:137]: Consider moving this into the above loop, and simply building `identifiersInRange` inside that loop, rather than collecting all identifiers and filtering later.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+
+        var methodsInRange = methodsInClass.Where(method => identifiersInRange
+                                    .Any(identifier => SymbolEqualityComparer.Default.Equals(identifier.Symbol, semanticModel.GetDeclaredSymbol(method))));
+
+        var fieldsInRange = fieldsInClass.Where(field => field.Declaration.Variables
+                                            .Any(variable => identifiersInRange
+                                                .Any(identifier => SymbolEqualityComparer.Default.Equals(identifier.Symbol, semanticModel.GetDeclaredSymbol(variable)))));
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:144]: This comment applies to the `methodsInRange` and `propertyInRange` calculation too, but this happens to be the worst offender: Consider how many times `GetDeclaredSymbol` is called in this calculation, for each specific value of `variable`.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+
+            pooledMethods.Add(new MethodSymbolicInfo
+            {
+                Name = method.Identifier.Text,
+                ReturnType = GetFullTypeName(method.ReturnType!, semanticModel),
+                ParameterTypes = parameterTypes
+            });
+        }
+
+        var expressionIdentifiersInRange = identifiersInRange.Where(i => i.Identifier.Ancestors().OfType<ExpressionStatementSyntax>().Any());
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:169]: Comment what this is finding? Why isn't this needed for methods? etc.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+            return false;
+        }
+
+        var baseTypes = classDeclaration.BaseList?.Types;
+        if (baseTypes is null)
+        {
+            return false;
+        }
+
+        foreach (var baseTypeSyntax in baseTypes)
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:208]: This loop seems unnecessary. I think it should be possible to get the symbol info for the class declaration itself (GetDeclaredSymbol I think) and then `InheritsFrom` could be called directly with that type symbol.

File: src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+        if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+        {
+            // Get the base name of the type, e.g., "List"
+            var typeName = namedTypeSymbol.Name;
+
+            // If there are type arguments, format them recursively
+            if (namedTypeSymbol.TypeArguments.Length > 0)
+            {
+                var typeArguments = string.Join(", ", namedTypeSymbol.TypeArguments.Select(FormatType));
+                return $"{typeName}<{typeArguments}>"; // Returning a formatted string seems hacky so might need to be revisited.
```
Comment (davidwengier) [inline src/LanguageServer/Protocol/ExternalAccess/Razor/GetSymbolicInfoHandler.cs:296]: Best not to do this manually, but rather use a call to `ToDisplayString()` that specifies an appropriate set of `SymbolDisplayFormat` values to create the desired result.
